specs/virtual_state/begin_tx.md

```diff
@@ -0,0 +1,106 @@
+# BEGIN_TX 
+
+## Procedure
+
+The `begin_tx` gadget performs preliminary checks on a transaction and its caller's account as well as further checks depending on the transaction type. 
+
+More particularly, for all transactions, we make sure that the account nonce is incremented by 1 if the tx is valid, and that the caller address is not null. We also verify that a transaction is set as invalid if and only if there is either not enough gas, the tx's nonce is not the same as caller's account previous nonce or the balance is insufficient.
+
+Depending whether the transaction is a Create, calls a precompile or calls an account with or without code, the checks as well as the number of lookups vary.
+
+## Constraints
+
+1. If this is the first transaction, check the transaction id is set to 1
+-> `tx_id = 1`
+2. Check the transaction's caller's address is not null
+-> `tx_caller_address != 0`
+3. If the transaction is valid, check the account nonce has been incremented correctly otherwise is the same
+-> `account_nonce - account_nonce_prev - 1 + is_tx_invalid = 0`
+4. Initilialise transactions's access list with caller's, callee's and coinbase's address
+5. Verify transfer with `tx_value` and `gas_fee` or 0s if tx is invalid
+6. Assert is_tx_invalid is correct
+-> `is_tx_invalid  - 1 - (1 - balance_not_enough) * (1 - gas_not_enough) * (is_nonce_valid) = 0`
```

Comment:
I've been thinking we shouldn't spell out all the arithmetization details in the spec. Unless we intentionally deploy some optimization tricks, otherwise we can abstract out the arithmetization details most of the time.
It would already be super helpful to mention just the three conditions that must be satisfied.
- sufficient balance: sender balance > value + fee
- sufficient gas: gas limit > gas used
- valid nonce: sender nonce = tx nonce

Reply:
Updated